crates/hotshot/src/traits/networking/push_cdn_network/definition/message_hook.rs

```diff
@@ -0,0 +1,577 @@
+#![allow(clippy::cast_sign_loss)]
+#![allow(clippy::cast_precision_loss)]
+
+use std::cmp;
+use std::hash::Hasher;
+use std::num::NonZeroUsize;
+use std::sync::atomic::{AtomicU32, Ordering};
+use std::sync::Arc;
+use std::time::{Duration, Instant};
+
+use anyhow::{Context, Result};
+use cdn_broker::reexports::def::hook::{HookResult, MessageHookDef};
+use cdn_broker::reexports::message::{Broadcast, Direct, Message as PushCdnMessage};
+use gcr::{Gcr, GcrRequestError};
+use lru::LruCache;
+use parking_lot::Mutex;
+use tracing::warn;
+use twox_hash::xxh3::Hash64;
+
+/// The type of message being processed. Is used downstream to determine
+/// which sample and average to use when processing a message.
+#[derive(Eq, PartialEq, Clone, Copy, Debug)]
+enum MessageType {
+    /// A broadcast message
+    Broadcast,
+    /// A direct message
+    Direct,
+}
+
+/// An average representing the number of bytes processed per period
+struct Average {
+    /// The number of bytes processed
+    num_bytes: Mutex<f64>,
+
+    /// The number of samples
+    num_samples: Mutex<u64>,
+
+    /// The last committed value
+    #[allow(clippy::struct_field_names)]
+    last_committed_average: AtomicU32,
+}
+
+impl Average {
+    /// Create a new `Average` instance
+    fn new() -> Self {
+        Self {
+            num_bytes: Mutex::new(0.0),
+            num_samples: Mutex::new(0),
+            last_committed_average: AtomicU32::new(0),
+        }
+    }
+
+    /// Add a sample to the average
+    fn add(&self, num: f64) {
+        let mut num_bytes = self.num_bytes.lock();
+        *num_bytes += num;
+
+        let mut num_samples = self.num_samples.lock();
+        *num_samples += 1;
+    }
+
+    /// Get the last committed average
+    fn get_last_committed_average(&self) -> u32 {
+        self.last_committed_average.load(Ordering::Relaxed)
+    }
+
+    /// Finalize the average, resetting the internal state and caching the new average.
+    /// Returns `true` if we committed a new average, `false` otherwise.
+    fn commit_if_necessary(&self, num_required_samples: usize) -> bool {
+        let mut num_samples = self.num_samples.lock();
+
+        // Return early if we don't have enough samples
+        if *num_samples < num_required_samples as u64 {
+            return false;
+        }
+
+        // Calculate the new average
+        let mut num_bytes = self.num_bytes.lock();
+        let new_average = (*num_bytes / *num_samples as f64) as u32;
+
+        // Reset the internal state
+        *num_bytes = 0.0;
+        *num_samples = 0;
+
+        // Cache the new average
+        self.last_committed_average
+            .store(new_average, Ordering::Relaxed);
+
+        // Return the new average
+        true
+    }
+}
+
+/// A sample representing the number of bytes processed per period
+#[derive(Clone)]
+pub struct Sample {
+    /// The number of bytes processed
+    num_bytes: f64,
+
+    /// The time of the last commit
+    last_committed_time: Instant,
+}
```

Comment:
This `Sample` is held locally for each CDN connection and is used to occasionally the global average